### PR TITLE
Re-architecture the "TaskList" entity in the backend

### PR DESCRIPTION
**Pull request by Atala**
Solves also #4007

### CODE REVIEW

File: js/app/dashboard/components/Tour.js [hunk 1]
```diff
@@ -91,7 +92,11 @@ const RenderEditNameForm = ({children, tour, isLoading}) => {
 }
 
 
-const Tour = ({ tour, draggableIndex }) => {
+const Tour = ({ tour, tourId, draggableIndex }) => {
+
+  if (!tour && tourId) {
+    tour = useSelector(state => selectTourById(state, tourId))
```
Comment (vladimir-8) [inline js/app/dashboard/components/Tour.js:98]: Hooks should not be called conditionally: https://react.dev/reference/rules/rules-of-hooks#only-call-hooks-at-the-top-level

File: js/app/dashboard/redux/actions.js [hunk 2]
```diff
@@ -292,63 +304,47 @@ export function importError(token, message) {
...
     const date = selectSelectedDate(state)
 
-    const url = window.Routing.generate('admin_task_list_modify', {
+    const url = window.Routing.generate('api_task_lists_set_items_item', {
       date: date.format('YYYY-MM-DD'),
       username,
     })
 
+    const { jwt } = getState()
+    const httpClient = createClient(dispatch)
```
Comment (vladimir-8) [inline js/app/dashboard/redux/actions.js:334]: Did you see another http client implementation in `user-data.js`? I think we should standardise on one implementation at some point

**Review by vladimir-8 — CHANGES_REQUESTED** [review]
`Request changes` is only due to the hook usage, all the rest are just suggestions

I guess with such amount of changes we should probably start releasing it with smaller instances; also checking if nothing breaks accidentally in the mobile app

File: src/Service/RouteOptimizer.php [hunk 3]
```diff
@@ -50,77 +54,95 @@ public function optimize(TaskCollection $taskCollection)
...
+        $items = $taskCollection->getItems();
+
+        foreach ($items as $item) {
+            if (null !== $item->getTour() && !in_array($item->getTour(), $tours, true)) {
+                $tours[] = $item->getTour();
+            } else if (null == $item->getTour()) {
+                $task = $item->getTask();
+
+                // FIXME : may not work as expected now that we allow to split deliveries pickup/dropoffs between riders
+                if (null !== $task->getDelivery() && !in_array($task->getDelivery(), $deliveries, true)) {
```
[the remark this answers is not in the document]
Comment (Atala) [reply src/Service/RouteOptimizer.php:119]: no, as
- `Delivery` is more info than tasks for the optimizer, it is an ordered set of tasks (better)

File: tests/AppBundle/Entity/TaskCollectionTest.php
[the remark this answers is not in the document]
Comment (Atala) [reply tests/AppBundle/Entity/TaskCollectionTest.php:1]: not yet

actually `TaskCollection`  (ordered set of tasks, which has a `distance` and `polyline`) is an abstract class/base class that has now two implementations `Delivery` and `Tours` (`TaskList` was but not anymore).

for your information we use doctrine inheritance mapping to translate this abstraction to the database-level.

(I am not really fond of this both on the code side or database side as  it adds complexity to the understanding of both the code and the codebase, but it is like this for now)

File: tests/AppBundle/Entity/TaskListTest.php [hunk 4]
```diff
@@ -16,21 +19,21 @@ public function setUp(): void
         $this->taskCollection->setCourier($user);
     }
 
-
+    // TODO : this should pass once we setup the listener
```
[the remark this answers is not in the document]
Comment (Atala) [reply tests/AppBundle/Entity/TaskListTest.php:22]: actually i removed the test ^^ and added comments in 143390905 to explain better

to explain this in few words: 
- for now web dispatch set the tasklist data AND the task data in TaskListManager->assign
- for now smartphone app dispatch sets only task data and this data is sync with tasklist in EntityChangeSetProcessor
- but i would really like to have both dispatch use the TaskList manager, and in this case tasklist->addTask and tasklist->removeTask have not use and can be removed

File: src/Action/TaskList/SetItems.php [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+
+final class SetItems
+{
+    public function __construct(
+        protected EntityManagerInterface $objectManager,
+        protected UserManager $userManager,
+        protected IriConverterInterface $iriConverter,
+        protected TaskListManager $taskListManager,
+        protected TaskListNormalizer $taskListNormalizer
+    )
```
[the remark this answers is not in the document]
Comment (Atala) [reply src/Action/TaskList/SetItems.php:25]: OK

File: js/app/dashboard/redux/actions.js (hunk 2 above)
Comment (Atala) [reply js/app/dashboard/redux/actions.js:334]: https://github.com/coopcycle/coopcycle-web/issues/4338

**Pull request merged**